### DISCUSSION TITLE
Adding new device option and inserting binary machine code into hal.executable.binary

### DISCUSSION THREAD

**Discussion by WoongQ** [iree-org/iree#12609] [Q&A]
Hello, I am attempting to add a new device option to the IREE compiler and runtime. Specifically, I would like to insert the binary machine code for our device, which is similar to .spv or PTX code, into hal.executable.binary. However, I have noticed that the hal.executable.binary operation contains [additional entry and end codes](https://github.com/openxla/iree/blob/main/compiler/src/iree/compiler/Dialect/HAL/Target/VulkanSPIRV/VulkanSPIRVTarget.cpp#L183) with serialized device codes, as seen in the [SerializeExecutable pass](https://github.com/openxla/iree/blob/main/compiler/src/iree/compiler/Dialect/HAL/Target/VulkanSPIRV/VulkanSPIRVTarget.cpp#L126).

As such, I would like to ask how the format of hal.executable.binary is structured, including the roles of the entry point and end codes. Additionally, I would like to know if there are any other considerations to keep in mind when inserting the new device code.

Thank you.

**Comment by benvanik** [comment]
The binary data is entirely runtime backend-specific - whatever the runtime side needs must be carried across. For SPIR-V/Vulkan we need to know how to map entry point ordinals into functions and subgroup sizes so we pack those along as well, while the CPU backend doesn't include any additional metadata as the executables are self-describing.
